More Efficient Ordering of Shard Upload Execution

original-brownbear opened the pull request:
Inspired by https://github.com/elastic/elasticsearch/pull/39657 and resolving the issue raised in that PR:

The problem currently is that we are uploading all segments in a shard sequentially per shard and only parallelize snapshotting across shards but not within shards.
This PR adjusts the logic towards parallel segment uploads: 
* Instead of starting all shard uploads in parallel and without any order, run segment uploads shard by shard.
* Only acquire index commit when it's actually required (after determining the current state of a shard in the repository, not before that).
   * @DaveCTurner pointed out that going sequentially in determining the segments to upload shard by shard increases the timespan between index commits. This is true for the case pf having more snapshot threads than shards to be snapshotted. On the other hand if there's more shards to be snapshotted on a node than snapshot threads it's the other way around. Currently, in this situation once the snapshot thread pool is fully occupied a shard will have to wait for another shard to completely finish its snapshot before taking the index commit (which could be many minutes!) while now the time span between the index commits is likely never more than a few seconds apart (biggest difference between two index commit times is basically equal to the time it takes all the shard folder listing operations to finish) 
* Release index commit asap (with this change we generally way reduce the amount of time we hold on to an index commit in cases of uploading multiple shards with multiple segments)

original-brownbear committed:
startt

original-brownbear committed:
simpler

original-brownbear committed:
start

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
fixed tests

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
snapshot pool for setting up new shards is fine after all

original-brownbear committed:
drier

original-brownbear committed:
drier

original-brownbear committed:
drier

original-brownbear committed:
drier

original-brownbear committed:
simpler

original-brownbear committed:
safer looking

original-brownbear committed:
some docs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
finished merge

original-brownbear committed:
bcl

original-brownbear committed:
simplify

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
simpler

ywelsch commented:
@tlrx can you have a look here?

tlrx commented:
@original-brownbear Do you think that this PR could be split into multiple smaller PRs? It is addressing several points (index commit acquiring/releasing, parallelization of uploads, factorization in a context class) and I must admit I have trouble to review this.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
shorter diff

original-brownbear commented:
@tlrx turns out this isn't so trivial to split up. Maybe to give you some background here take a look at https://github.com/elastic/elasticsearch/pull/42213 where we actively removed the index shard dependency from the repository.
Making a standalone change with the goal of 

>  I'd start with a simple change about where the indexcommit is acquired (after listing blobs) and released.

kind of tricky. So I'd have to bring the shard snapshot context abstraction back in isolation to achieve this I'm afraid.

The problem in general is, that the specific behavior change here (the shard upload ordering) is very small and merely a simple reordering of two for-loops being adjusted from parallel to non parallel and vice-versa. All the changes around it are merely there to enable doing this in a non-blocking manner (it would be a literal 10 line change if I could block inside the `snapshotShard` method).
The other thing is, moving the place where we acquire the shard index commit without the changes to parallelizing file uploads would actually make the problem addressed here worse I think. Now we'd wait even longer (because the whole blob listing IO is added to things ... even though it won't matter much practically) between acquiring two index commits and wouldn't get any faster release of those commits either.

Maybe it would be more helpful if I were to create a blocking version of just the shard upload reordering to make things clearer. We wouldn't be able to merge that as it would break the `SnapshotResiliencyTests` that rely on non of the snapshot methods blocking but I think that might make this easier to follow?
Another option would be to maybe just talk this through live next week? :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear commented:
Please hold off on reviewing this one, after the latest merge of `master` into this branch we seem to have picked up some conflict with SLM behaviour here and this change makes the following test run:

```
./gradlew :x-pack:plugin:ilm:qa:multi-node:integTestRunner -Dtests.seed=39DDC190C49A92E8 -Dtests.security.manager=true -Dtests.locale=kab-DZ -Dtests.timezone=Africa/Bamako -Dcompiler.java=12 -Druntime.java=11
```

flaky -> have to track this down first.

original-brownbear commented:
Sort of false alarm here, the test failures are a result of https://github.com/elastic/elasticsearch/issues/21759 (the SLM test that failed uses a rate limit of 1 byte per second so if parallel writers hit that limit they can way overshoot that limit an make all the SNAPSHOT threads wait. I'll open a separate PR to improve the situation behind aborting snapshots, doing that here would way overcomplicate this PR.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear commented:
Jenkins run elasticsearch-ci/2

original-brownbear commented:
Jenkins run elasticsearch-ci/2

tlrx commented on server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java:
```diff
@@ -0,0 +1,162 @@
+/**
+ * Manages the {@link IndexCommit} associated with the shard snapshot as well as its {@link IndexShardSnapshotStatus} instance.
+ */
+public class ShardSnapshotContext {
```
At first sight I liked this new class, but after iterating over this PR few times I like it a bit less. What I don't like here is that it adds another place where the `IndexShardSnapshotStatus` is modified (and kind of hide these modifications), and I also feel like it adds extra complexity.

Instead, I think we could wrap the IndexCommit into a AbstractRefCounted, increment it before triggering a file snapshot, decrement it once the snapshot of the file is done and let the IndexCommit be released itself once all file snapshots are done.

original-brownbear replied:
> Instead, I think we could wrap the IndexCommit into a AbstractRefCounted, increment it before triggering a file snapshot, decrement it once the snapshot of the file is done and let the IndexCommit be released itself once all file snapshots are done.

That was my plan A at first. But then I started not liking that so much because:
* It makes the signature `org.elasticsearch.repositories.blobstore.BlobStoreRepository#snapshotShard` even more complex than it already is
* While it makes the state transitions of `IndexShardSnapshotStatus` a little more obvious now, it hides the connection between that ref counted instance, the index commit and the status doesn't it?

... Maybe it would be better to just flatten all this logic into `IndexShardSnapshotStatus` (or the logic of the shard status tracking into `ShardSnapshotContext`)? :)

The other thing about me adding this was that we will need to keep track of more state for each shard upload when fixing/optimizing the situation of unchanged shards in https://github.com/elastic/elasticsearch/issues/45736 ... my thinking was that if we change the signature like it was done here, we could do without yet another subsequent signature change across all repository implementations when dealing with that and simply add the necessary logic to convey an unchanged shard and its generation via `ShardSnapshotContext`?

tlrx replied:
It looks to me that we are trying to address different issues in this PR:
a) more efficient way to snapshot shard files
b) better visibility and better tracking of shard status and resources releases
c) prepare the work for #45736

I think this PR should be just about point a). I'm not saying that b) and c) are not valid and I'm pretty sure they deserve their own PRs :) but for now I don't think they should be an argument to introduce more complexity for what we want to solve here.

What we could do (sorry again if I wasn't clear before) is to acquire an index commit reference in  `SnapshotShardsService.snapshot()` like it was before, and ensure that the reference is released when the shard snapshot listener gets called. Then the listener is passed down to the `BlobStoreRepository#snapshotShard()`  which makes sure to call it through the group files listener. This way we don't touch how status is tracked and we don't introduce another complex object (which again, could be the subject of a dedicated PR). WDYT?

original-brownbear replied:
@tlrx I think my main point above all the others was `c)` (sorry for not making that clear) and the fact that we would have to change the signature of `snapshotShard` twice (especially when those versions of the signature may be spread out across releases).

I think other than that, I just liked this approach because it kept the logic a little less spread out. But if the above point can't warm your heart to this change I can probably do what you're asking for pretty easily :)

original-brownbear replied:
Done :) Now we're doing all the resource handling in the callback and are still creating the commit ref in the snapshot shards service like we did before.
Thanks for pushing on this ... turns out this massively simplified the change for the time being (200+ LoC changes saved :)) ... not sure why I did it so complicated the first time around.

tlrx reviewed (commented):
Sorry for the delay it took me to look at this. I left few comments I'd like to see addressed.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -909,132 +916,139 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
+                    }
+                    context.finish(threadPool.absoluteTimeInMillis());
+                }, onFailure), indexIncrementalFileCount + 1);
+            filesListener.onResponse(null);
```
I figured this was a nicer way of handling the special case of `0` files than having a condition for the case of zero files and then using another level of indirection for `onResponse` in the grouped listener. Apparently not and conditional wins? :D

tlrx replied:
To be honest, I just find this confusing :(

original-brownbear replied:
I will adjust this then once we settle on the other open point :)

original-brownbear replied:
Kileld it -> now it's just a conditional here :)

original-brownbear committed:
CR: move try-catch into snapshot method

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -909,132 +916,139 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
+                    @Override
+                    protected void doRun() {
+                        try {
+                            snapshotFile(snapshotFileInfo, indexId, shardId, snapshotId, snapshotStatus, store);
```
@tlrx yea that sounds like a reasonable approach for sure. The only issue with that, that I can see is this snippet in `snapshotFile`:

```
        } catch (Exception t) {
            failStoreIfCorrupted(store, t);
            snapshotStatus.addProcessedFile(0);
            throw t;
```

-> it is a little tricky to convert this into what you propose because we need the error but also the effect of a `0` return. Maybe we should make this nicer in a follow up? (also keeps the change complexity down a little to not change the abort mechanics here?)

tlrx replied:
The usage of snapshotStatus.addProcessedFile(0); should be the responsability of the caller, which should catch any exception.

+1 for a follow up

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -300,44 +296,36 @@ private void startNewSnapshots(SnapshotsInProgress snapshotsInProgress) {
+                    snapshot(indicesService.indexServiceSafe(
+                        shardId.getIndex()).getShardOrNull(shardId.id()), snapshot, indexId, snapshotStatus, listener);
+                } catch (Exception e) {
+                    listener.onFailure(e);
```
Done :)

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -364,21 +352,9 @@ private void snapshot(final IndexShard indexShard, final Snapshot snapshot, fina
-        } catch (Exception e) {
-            throw new IndexShardSnapshotFailedException(shardId, "Failed to snapshot", e);
-        }
+        // we flush first to make sure we get the latest writes snapshotted
```
I just removed it, it doesn't apply in any way now.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -909,132 +916,139 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
+                            snapshotFile(snapshotFileInfo, indexId, shardId, snapshotId, snapshotStatus, store);
+                            filesListener.onResponse(null);
+                        } catch (IOException e) {
+                            throw new IndexShardSnapshotFailedException(shardId, "Failed to perform snapshot (index files)", e);
```
True, but maybe we should change it in another PR to not mix that change into this refactoring? (It was the same message before this change)

original-brownbear commented:
@tlrx no worries with the delay, I know this is a nasty one -> thanks for taking a look! :) Tried addressing all the points you raised, let me know what you think.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
simplify signature

original-brownbear committed:
nicer looking

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
fix noisy empty line

original-brownbear committed:
stop being clever aobut group action listener

original-brownbear commented:
@tlrx I've now cleaned up all the things we talked about. This should be ready for another round of reviewing (of a much shorter diff :)).

tlrx commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -226,6 +227,37 @@ public void onFailure(Exception e) {
 
+    /**
+     * Wraps a given listener and returns a new listener which executes the provided {@code runBefore}
+     * callback before the listener is notified ia either {@code #onResponse} or {@code #onFailure}.
```
typo: ia

tlrx commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -226,6 +227,37 @@ public void onFailure(Exception e) {
+     * If the callback throws an exception then it will be passed to the listener's {@code #onFailure} and its {@code #onResponse} will
+     * not be executed.
+     */
+    static <Response> ActionListener<Response> runBefore(ActionListener<Response> delegate, CheckedRunnable<?> runBefore) {
```
Can you add a test for this?

original-brownbear replied:
Sure, I just added the same one we have for the after hook here as well :)

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -346,38 +330,36 @@ public void onAfter() {
+            final Repository repository = repositoriesService.repository(snapshot.getRepository());
+            Engine.IndexCommitRef snapshotRef = null;
+            try {
+                snapshotRef = indexShard.acquireLastIndexCommit(true);
```
Maybe we could restore the comment about why we flush here?

original-brownbear replied:
Right, brought it back :)

tlrx commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -907,132 +914,144 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
-            } catch (IOException e) {
-                throw new IndexShardSnapshotFailedException(shardId, "Failed to write commit point", e);
-            }
+            final Runnable afterUploads = () -> {
```
I don't really like this `Runnable`, maybe we should use `StepListener`s instead?

I'm thinking of something like:
```java
final StepListener<Void> snapshotDone = new StepListener<>();
snapshotDone.whenComplete(listener::onResponse, onFailure); // snapshot is done, call parent listener

final StepListener<Void> snapshotFinalization = new StepListener<>();
snapshotFinalization.whenComplete(v -> {
    ... what is in the runnable...
    snapshotDone.onResponse(null);
}, snapshotDone::onFailre);

if (indexIncrementalFileCount == 0) {
    snapshotFinalization.onResponse(null);
} else {
    ... use the GroupedActionListener with a wrapped snapshotFinalization...
}
```

WDYT?

original-brownbear replied:
I like it :) Used a step listener the way you suggested now (with one naming change).

tlrx replied:
This is nice :+1:

tlrx reviewed (commented):
This looks great, thanks @original-brownbear! I prefer it this way.

I've left a suggestion about using StepListener, otherwise this is getting close.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into smarter-ordering-shard-upload

original-brownbear committed:
CR: step listeners, test, comment,typo

original-brownbear commented:
Thanks @tlrx addressed all your points now I think :)

tlrx reviewed (approved):
LGTM

Thank you for considering my suggestions, I really like it now :) I think we should hold on the backport of this until 7.4 is cut.

original-brownbear commented:
Thanks so much @tlrx! + def. this isn't `7.4` material :)
